app/jobs/purge_soft_deleted_records_job.rb

```diff
@@ -0,0 +1,112 @@
+          models.each do |model|
+            model.unscoped do
+              process_model(model, data_source: data_source)
+            end
+          end
+        end
+      end
+    end
+    @total_deleted
+  end
+
+  protected
+
+  def data_sources
+    GrdaWarehouse::DataSource
+  end
+
+  # client-related warehouse models
+  def warehouse_models
+    [
+      # enrollment-dependent
+      GrdaWarehouse::Hud::Assessment,
+      GrdaWarehouse::Hud::AssessmentQuestion,
+      GrdaWarehouse::Hud::AssessmentResult,
+      GrdaWarehouse::Hud::CurrentLivingSituation,
+      GrdaWarehouse::Hud::Disability,
+      GrdaWarehouse::Hud::EmploymentEducation,
+      GrdaWarehouse::Hud::Event,
+      GrdaWarehouse::Hud::Exit,
+      GrdaWarehouse::Hud::HealthAndDv,
+      GrdaWarehouse::Hud::IncomeBenefit,
+      GrdaWarehouse::Hud::Service,
+      GrdaWarehouse::Hud::YouthEducationStatus,
+      Hmis::Hud::CustomAssessment,
+      Hmis::Hud::CustomCaseNote,
+      Hmis::Hud::CustomClientAddress,
+      Hmis::Hud::CustomClientContactPoint,
+      Hmis::Hud::CustomClientName,
+      Hmis::Hud::CustomDataElement,
+      # purge these last
+      GrdaWarehouse::Hud::Enrollment,
+      GrdaWarehouse::Hud::Client,
+    ]
+  end
+
+  # tables with FK relationships need to be deleted. Choosing to leave other dangling references to client
+  def client_dependents(client_scope)
+    # double check that these are the same table before we start deleting records with that assumption
+    raise unless GrdaWarehouse::Hud::Client.table_name == Hmis::Hud::Client.table_name
+    raise unless GrdaWarehouse::Hud::Client.connection.current_database == Hmis::Hud::Client.connection.current_database
+
+    rhm_t = HmisExternalApis::AcHmis::ReferralHouseholdMember.arel_table
+    [
+      GrdaWarehouse::WarehouseClient.joins(:destination).merge(client_scope),
+      GrdaWarehouse::WarehouseClient.joins(:source).merge(client_scope),
+      GrdaWarehouse::WarehouseClientsProcessed.joins(:client).merge(client_scope),
+      HmisExternalApis::AcHmis::ReferralHouseholdMember.where(rhm_t[:client_id].in(client_scope.pluck(:id))),
+    ]
+  end
+
+  def with_lock(&block)
+    lock_name = self.class.name.demodulize
+    GrdaWarehouseBase.with_advisory_lock(lock_name, timeout_seconds: 0, &block)
+  end
+
+  def process_model(model, data_source:)
+    arel = model.arel_table
+    paranoia_col = arel[model.paranoia_column.to_sym]
+    scope = model.
+      where(data_source: data_source).
+      where(paranoia_col.lt(@retain_at))
+
+    scope.in_batches.each do |batch|
+      model.transaction do
+        if model == GrdaWarehouse::Hud::Client
+          client_dependents(batch).each do |dependent_scope|
+            check_max_deleted(dependent_scope.size)
+            dependent_scope.delete_all unless @dry_run
+          end
+        end
+
+        check_max_deleted(batch.size)
+        batch.delete_all unless @dry_run
+      end
+    end
+  end
```

Comment:
I'm probably missing something here, but couldn't we skip the batching and just do something like the following?
```
num_deleted = scope.limit(@max_deleted).delete_all
check_max_deleted(num_deleted)
```

As long as you loop over the dependent items first/client last, the client records should be safe to delete once everything else is gone.

Reply:
My thought was that batches result in smaller individual transactions that should be less load on the db (smaller memory peak usage but a longer run-time).  It also makes the dependent joins a little easier. I could be convinced otherwise though

Reply:
I kicked up the batch size to 5k, we can see how it runs on QA